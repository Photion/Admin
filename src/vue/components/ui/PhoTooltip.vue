<template>
  <div class="wrapper relative">
    <div class="inline-block" @mouseout="show = false" @mouseover="show = true">
      <slot ></slot>
    </div>
    <transition name="slide" @enter="setEnterDelay" @leave="setLeaveDelay" >
      <div v-if="show" class="tooltip bg-gray-800 text-white rounded-sm text-center px-2 py-1">
        <span v-if="text" class="text-xs tracking-wide">{{text}}</span>
        <div v-else>
          <slot name="content"></slot>
        </div>
      </div>
    </transition>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue';

export default defineComponent({
  name: 'PhoTooltip',
  props: {
    text: {
      type: String,
      required: false,
    },
    delayIn: {
      type: Number,
      default: 1000,
    },
    delayOut: {
      type: Number,
      default: 100,
    },
  },
  setup(props) {
    const transitionDelayIn = computed<string>(() => `${props.delayIn}ms`);
    const transitionDelayOut = computed<string>(() => `${props.delayOut}ms`);

    return {
      transitionDelayIn,
      transitionDelayOut,
    };
  },
  data() {
    return {
      show: false,
    };
  },
  methods: {
    setEnterDelay(el: HTMLElement, _done: boolean) {
      el.style.transitionDelay = this.transitionDelayIn ;
    },
    setLeaveDelay(el: HTMLElement, _done: boolean) {
      el.style.transitionDelay = this.transitionDelayOut ;
    },
  },
});
</script>

<style lang="scss" scoped>
.wrapper {
  width: fit-content;
}

.tooltip {
  position: absolute;
  max-width: 12rem;
  top: 130%;

  &::after {
    content: "";
    position: absolute;
    bottom: 100%;
    left: 50%;
    margin-left: -5px;
    border-width: 5px;
    border-style: solid;
    //noinspection CssInvalidFunction
    border-color: transparent transparent theme('colors.gray.800') transparent;
  }
}

.slide-enter-active {
  transition-duration: 150ms;
}

.slide-leave-active {
  transition-duration: 150ms;
}

.slide-enter-from {
  opacity: 0;
  transform: translateY(-20%);
}

.slide-enter-to , .slide-leave-from {
  opacity: 1;
  transform: translateY(0);
}

.slide-leave-to {
  opacity: 0;
  transform: translateY(-20%);
}
</style>