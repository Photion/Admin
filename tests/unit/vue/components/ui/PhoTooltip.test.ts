import { mount } from '@vue/test-utils';

import PhoTooltip from '~/src/vue/components/ui/PhoTooltip.vue';

const getComponent = <T, Y>(props: T, slots?: Y) => {
  return mount(PhoTooltip, {
    props: {
      name: 'pho-tooltip-comp',
      uuid: 'my-uuid',
      ...props,
    },
    slots: {
      default: 'Tooltip Trigger',
      ...slots,
    },
  });
};

describe('unit.vue.components.ui.PhoTooltip', () => {

  const positions = ['top', 'right', 'bottom', 'left'];

  beforeEach(() => jest.useFakeTimers());

  it('Mounts', () => {
    const wrapper = getComponent({});
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('On mouse enter', () => {
    positions.forEach((position) => {
      it(`Renders the text on '${position}'`, async () => {
        const wrapper = getComponent({ text: 'Some Tooltip Text', position });
        await wrapper.trigger('mouseenter');
        await jest.runOnlyPendingTimers();
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    positions.forEach((position) => {
      it(`Renders the content slot on '${position}'`, async () => {
        const content = '<div>Tooltip with <strong>strong</strong> html content</div>';
        const wrapper = getComponent({}, { content, position });
        await wrapper.trigger('mouseenter');
        await jest.runOnlyPendingTimers();
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    it('Sets the attributes when it is shown', async () => {
      const wrapper = getComponent({});
      await wrapper.trigger('mouseenter');
      await jest.runAllTimers();
      const attributes = wrapper.get('[cy]').attributes();

      expect(attributes).toMatchSnapshot();
    });

  });

  describe('On mouse leave', () => {
    it('Hides the tooltip', async () => {
      const wrapper = getComponent({ text: 'Some Tooltip Text' });
      await wrapper.trigger('mouseenter');
      await jest.runAllTimers();
      await wrapper.trigger('mouseleave');
      await jest.runAllTimers();
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('On mouse click', () => {
    describe('With one click', () => {
      it('Shows the tooltip', async () => {
        const wrapper = getComponent({ text: 'Some Tooltip Text' });
        await wrapper.trigger('click');
        await jest.runAllTimers();
        expect(wrapper.element).toMatchSnapshot();
      });

      it('Overrides mouseenter', async () => {
        const wrapper = getComponent({ text: 'Some Tooltip Text' });
        await wrapper.trigger('mouseenter');
        await wrapper.trigger('click');
        await jest.runAllTimers();
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('With two clicks', () => {
      it('Hides the tooltip', async () => {
        const wrapper = getComponent({ text: 'Some Tooltip Text' });
        await wrapper.trigger('click');
        await jest.runAllTimers();
        await wrapper.trigger('click');
        await jest.runAllTimers();
        expect(wrapper.element).toMatchSnapshot();
      });
    });

  });

});
