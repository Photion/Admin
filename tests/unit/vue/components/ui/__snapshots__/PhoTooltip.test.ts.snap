// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit.vue.components.ui.PhoTooltip Mounts 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse click With one click Overrides mouseenter 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <div
      ci="tooltip:pho-tooltip-comp:tooltip"
      class="tooltip absolute bg-gray-800 text-white rounded-sm text-center px-2 py-1 bottom"
      id="tooltip:pho-tooltip-comp:my-uuid:tooltip"
      style="--transition-delay-in: 700ms; --transition-delay-out: 300ms;"
    >
      
      <div
        class="text-xs tracking-wide"
      >
        Some Tooltip Text
      </div>
      
    </div>
  </transition-stub>
</div>
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse click With one click Shows the tooltip 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <div
      ci="tooltip:pho-tooltip-comp:tooltip"
      class="tooltip absolute bg-gray-800 text-white rounded-sm text-center px-2 py-1 bottom"
      id="tooltip:pho-tooltip-comp:my-uuid:tooltip"
      style="--transition-delay-in: 700ms; --transition-delay-out: 300ms;"
    >
      
      <div
        class="text-xs tracking-wide"
      >
        Some Tooltip Text
      </div>
      
    </div>
  </transition-stub>
</div>
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse click With two clicks Hides the tooltip 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse enter Renders the content slot on 'bottom' 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <div
      ci="tooltip:pho-tooltip-comp:tooltip"
      class="tooltip absolute bg-gray-800 text-white rounded-sm text-center px-2 py-1 bottom"
      id="tooltip:pho-tooltip-comp:my-uuid:tooltip"
      style="--transition-delay-in: 700ms; --transition-delay-out: 300ms;"
    >
      
      
      <div>
        Tooltip with 
        <strong>
          strong
        </strong>
         html content
      </div>
      
      
    </div>
  </transition-stub>
</div>
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse enter Renders the content slot on 'left' 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <div
      ci="tooltip:pho-tooltip-comp:tooltip"
      class="tooltip absolute bg-gray-800 text-white rounded-sm text-center px-2 py-1 bottom"
      id="tooltip:pho-tooltip-comp:my-uuid:tooltip"
      style="--transition-delay-in: 700ms; --transition-delay-out: 300ms;"
    >
      
      
      <div>
        Tooltip with 
        <strong>
          strong
        </strong>
         html content
      </div>
      
      
    </div>
  </transition-stub>
</div>
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse enter Renders the content slot on 'right' 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <div
      ci="tooltip:pho-tooltip-comp:tooltip"
      class="tooltip absolute bg-gray-800 text-white rounded-sm text-center px-2 py-1 bottom"
      id="tooltip:pho-tooltip-comp:my-uuid:tooltip"
      style="--transition-delay-in: 700ms; --transition-delay-out: 300ms;"
    >
      
      
      <div>
        Tooltip with 
        <strong>
          strong
        </strong>
         html content
      </div>
      
      
    </div>
  </transition-stub>
</div>
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse enter Renders the content slot on 'top' 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <div
      ci="tooltip:pho-tooltip-comp:tooltip"
      class="tooltip absolute bg-gray-800 text-white rounded-sm text-center px-2 py-1 bottom"
      id="tooltip:pho-tooltip-comp:my-uuid:tooltip"
      style="--transition-delay-in: 700ms; --transition-delay-out: 300ms;"
    >
      
      
      <div>
        Tooltip with 
        <strong>
          strong
        </strong>
         html content
      </div>
      
      
    </div>
  </transition-stub>
</div>
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse enter Renders the text on 'bottom' 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <div
      ci="tooltip:pho-tooltip-comp:tooltip"
      class="tooltip absolute bg-gray-800 text-white rounded-sm text-center px-2 py-1 bottom"
      id="tooltip:pho-tooltip-comp:my-uuid:tooltip"
      style="--transition-delay-in: 700ms; --transition-delay-out: 300ms;"
    >
      
      <div
        class="text-xs tracking-wide"
      >
        Some Tooltip Text
      </div>
      
    </div>
  </transition-stub>
</div>
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse enter Renders the text on 'left' 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <div
      ci="tooltip:pho-tooltip-comp:tooltip"
      class="tooltip absolute bg-gray-800 text-white rounded-sm text-center px-2 py-1 left"
      id="tooltip:pho-tooltip-comp:my-uuid:tooltip"
      style="--transition-delay-in: 700ms; --transition-delay-out: 300ms;"
    >
      
      <div
        class="text-xs tracking-wide"
      >
        Some Tooltip Text
      </div>
      
    </div>
  </transition-stub>
</div>
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse enter Renders the text on 'right' 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <div
      ci="tooltip:pho-tooltip-comp:tooltip"
      class="tooltip absolute bg-gray-800 text-white rounded-sm text-center px-2 py-1 right"
      id="tooltip:pho-tooltip-comp:my-uuid:tooltip"
      style="--transition-delay-in: 700ms; --transition-delay-out: 300ms;"
    >
      
      <div
        class="text-xs tracking-wide"
      >
        Some Tooltip Text
      </div>
      
    </div>
  </transition-stub>
</div>
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse enter Renders the text on 'top' 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <div
      ci="tooltip:pho-tooltip-comp:tooltip"
      class="tooltip absolute bg-gray-800 text-white rounded-sm text-center px-2 py-1 top"
      id="tooltip:pho-tooltip-comp:my-uuid:tooltip"
      style="--transition-delay-in: 700ms; --transition-delay-out: 300ms;"
    >
      
      <div
        class="text-xs tracking-wide"
      >
        Some Tooltip Text
      </div>
      
    </div>
  </transition-stub>
</div>
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse enter Sets the attributes when it is shown 1`] = `
Object {
  "class": "wrapper relative",
  "cy": "tooltip:pho-tooltip-comp",
  "id": "tooltip:pho-tooltip-comp:my-uuid",
}
`;

exports[`unit.vue.components.ui.PhoTooltip On mouse leave Hides the tooltip 1`] = `
<div
  class="wrapper relative"
  cy="tooltip:pho-tooltip-comp"
  id="tooltip:pho-tooltip-comp:my-uuid"
>
  <div
    class="inline-block w-full"
  >
    
    
    Tooltip Trigger
    
    
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
`;
